### PR TITLE
Correctly run all pending tasks for EmbeddedChannel when the Channel …

### PR DESCRIPTION
**Pull request by normanmaurer**
…was closed.

Motivation:

When a user called ctx.close() and used the EmbeddedChannel we did not correctly run all pending tasks which means channelInactive was never called.

Modifications:

Ensure we run all pending tasks after all operations that may change the Channel
 state and are part of the Channel.Unsafe impl.

Result:

Fixes [#6894].

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin can you check https://github.com/netty/netty/pull/6897 ?

File: transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java [hunk 1]
```diff
@@ -697,6 +697,13 @@ protected void handleInboundMessage(Object msg) {
         public void connect(SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise) {
             safeSetSuccess(promise);
         }
+
+        @Override
+        protected void afterClose() {
+            // Ensure we rull al pending tasks when the channel is closed.
```
Comment (trustin) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java]: run all

File: transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java [hunk 2]
```diff
@@ -541,6 +542,25 @@ protected void handleOutboundMessage(Object msg) {
         }
     }
 
+    @Test(timeout = 2000)
```
Comment (trustin) [inline transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java]: Does this have to be this short? Seems like it wants flakiness on a slow machine.

**Comment by normanmaurer** [comment]
@trustin thanks.. .fixed your nits

File: transport/src/main/java/io/netty/channel/AbstractChannel.java [hunk 3]
```diff
@@ -678,6 +682,7 @@ public void run() {
                     fireChannelInactiveAndDeregister(wasActive);
                 }
             }
+            afterClose();
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: can you add a comment as to why this cannot be accomplished in `doClose()`? It seems strange that this operation is done synchronously and the `doClose()` call may use `closeExecutor`.

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: yep will do...

Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannel.java]: let me think a bit more about this bugfix

**Comment by normanmaurer** [comment]
@Scottmitch PTAL again... I think this is better

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (8adb30bbe2785094c92d085757cb1fdc2c1edec9) and 4.0 (98140cb98f).

Thanks again @Scottmitch for the comments and so let me think more about this and come up with a better solution

**Pull request closed without merging**